HIT L2 - add summed flux algorithm

vmartinez-cu opened the pull request:
This PR initiates L2 processing by implementing the following:

- Adding a file to handle L2 processing
- Implementing an algorithm to create an L2 summed flux dataset in the new file 
- Adding ancillary files needed for L2 products
- Adding handling for L2 products in cli.py and the attribute yaml files

Closes #1073 
Closes #1071 

## New Files
- Added new folder with ancillary files
   - imap_hit_l1b-to-l2-summed-factors-20250219_v002.csv

## Deleted Files
- sci_sample_raw1.csv
   - Deleted validation file since it's been replaced by an updated version in a previous PR

## Updated Files
- imap_hit_global_cdf_attrs.yaml
   - Added L2 global attributes
- imap_hit_l2_variable_attrs.yaml
   - Added yaml file for L2 variables. Mostly empty for now.
- cli.py
   - Added handling for L2 products
- hit/l1b/constants.py
   - Renamed particles used in both L1B and L2 products
- hit_l1b.py
   - Renamed particles used in both L1B and L2 products
- test_hit_l1b.py
   - Renamed particles used in both L1B and L2 products
- hit_l2.py
   - Added new file to process L2 data
   - Added algorithm for summed flux product
- test_hit_l2.py
   - Added test for new L2 functions

vmartinez-cu committed:
WIP - create L2 data processing file. Implement summed flux algorithm

vmartinez-cu committed:
Update particle names to lower case abbreviated names (i.e. h instead of hydrogen)

vmartinez-cu committed:
Optimize function by using vectorization

vmartinez-cu committed:
Change input data type from dict to xarray dataset and fix typo in variable name

vmartinez-cu committed:
Add unit test file for L2 processing and add test for summed flux processing

vmartinez-cu committed:
Add attribute manager for L2 products

vmartinez-cu committed:
Add L2 processing for HIT

vmartinez-cu committed:
Clean up comments and remove testing code

vmartinez-cu committed:
Clean up test file

nkerman committed:
Reorder ultra map code dimensions -> (epoch, energy, az, el) (#1414)

vmartinez-cu committed:
Add ancillary files for L2 processing

tech3371 commented on imap_processing/cli.py:
```diff
@@ -583,17 +584,31 @@ def do_processing(self, dependencies: list) -> list[xr.Dataset]:
+                raise ValueError(
+                    f"Unexpected dependencies found for HIT L1b:"
+                    f"{dependencies}. Expected only two dependencies."
+                )
```
```suggestion
                raise ValueError(
                    f"Unexpected dependencies found for HIT L1b:"
                    f"{dependencies}. Expected only one dependency."
                )
```

tech3371 commented on imap_processing/cli.py:
```diff
@@ -583,17 +584,31 @@ def do_processing(self, dependencies: list) -> list[xr.Dataset]:
+                raise ValueError(
+                    f"Unexpected dependencies found for HIT L2:"
+                    f"{dependencies}. Expected only two dependencies."
+                )
```
```suggestion
                raise ValueError(
                    f"Unexpected dependencies found for HIT L2:"
                    f"{dependencies}. Expected only one dependency."
                )
```

tech3371 commented on imap_processing/hit/l2/hit_l2.py:
```diff
@@ -0,0 +1,163 @@
+
+    # Logical sources for the three L2 science products.
+    # TODO: add logical sources for other l2 products once processing functions
+    #  are written. "imap_hit_l2_standard-fluxes", "imap_hit_l2_sectored-fluxes"
```
Good to know

tech3371 commented on imap_processing/hit/l2/hit_l2.py:
```diff
@@ -0,0 +1,163 @@
+        logger.info("HIT L2 summed flux dataset created")
+
+    # Update attributes and dimensions
+    for dataset, logical_source in zip(l2_datasets, logical_sources):
```
Will the `logical_sources` and `l2_datasets` will always correctly map? I am wondering what would happen if one of those function return empty or something. I may be overthinking it. I am wondering if you want to do dictionary instead to be safe. Eg.
```
{
"imap_hit_l2_summed-fluxes": <it's xr.dataset returned from process_summed_flux_data()>,
so on for the other two in the future.
}
```

tech3371 commented on imap_processing/hit/l2/hit_l2.py:
```diff
@@ -0,0 +1,163 @@
+                    for key, value in attr_mgr.get_variable_attributes(field).items()
+                    if "DEPEND" in key
+                }
+                dataset[field].attrs = attr_mgr.get_variable_attributes(field)
```
I had to set `check_schema=False` on dimension's attrs or my CDF fails to load on SKTEditor. Something to be aware of for your future work.

vmartinez-cu replied:
Good to know. Do you know why the dimension attrs fail in SKTEditor?

tech3371 replied:
Because SAMMI adds additional stuff that dimension attrs doesn't need. That's why we had to add that check to be false so that it only adds what we want.

laspsandoval committed:
Ultra l1b - phi theta corrections (#1405)

* phi theta corrections

tech3371 commented on imap_processing/hit/l2/hit_l2.py:
```diff
@@ -0,0 +1,163 @@
+    xr.Dataset
+        The processed L2 summed flux dataset.
+    """
+    # TODO:
+    #  - determine where to pull ancillary data. Storing it locally for now
+    #  - add check for dynamic_threshold_state to determine which ancillary table to use
+    #    after additional ancillary files are provided
```
I am also waiting on this too. I will let you and others know when this is resolved.

tech3371 commented on imap_processing/hit/l2/hit_l2.py:
```diff
@@ -0,0 +1,163 @@
+                flux_factors = var_anc_data.set_index(
+                    var_anc_data["lower energy (mev)"].astype(np.float32)
+                ).loc[energy_min]
+                delta_e_factor = flux_factors["delta e (mev)"].values
+                geometry_factor = flux_factors["geometry factor (cm2 sr)"].values
+                efficiency = flux_factors["efficiency"].values
```
I hope they don't changes these column names :)

tech3371 commented on imap_processing/hit/l2/hit_l2.py:
```diff
@@ -0,0 +1,163 @@
+                efficiency = flux_factors["efficiency"].values
+                b = flux_factors["b"].values
+
+                # Calculate the summed flux for this energy bin
```
Thank you for great comments. It was helpful to follow the code!

tech3371 reviewed (approved):
Non blocking suggestions. Otheriwse, LGTM

tmplummer committed:
Add repoint module, test coverage, and general conftest fixture for g… (#1415)

* Add repoint module, test coverage, and general conftest fixture for generating fake repoint data on the fly.

* Update imap_processing/spice/repoint.py

Co-authored-by: Greg Lucas <greg.m.lucas@gmail.com>

* Improve docstrings
Add optional argument to seed the starting repoint_id number

---------

Co-authored-by: Greg Lucas <greg.m.lucas@gmail.com>

greglucas commented on imap_processing/hit/ancillary/imap_hit_l1b-to-l2-sector-factors_20250219_v002.csv:
Are these files going to be stored in the ancillary files in s3 later or are they static and not expecting to be changed?

Edit: Looks like they are anticipated to be in s3 later after reading later comments. A few other questions:
* Are you only using the summed one for now? Maybe only add that one now and add others later in the future PRs when you use them.
* Why is this file formatted differently than the others?
* You could look into mocking these tables similar to what we are doing for the spin/repointing dataframes. Make some idealized numbers for geometry/efficiency factors so you can assert against your mocked data rather than these potentially varying files.

vmartinez-cu replied:
- I'm only using the summed ancillary file for now, so I'll remove the others
- Are you referring to the format of the ancillary files or a different file?
- I'll look into mocking the tables when I validate the data in a separate PR. Would mocking the tables be better than using the test ancillary files HIT provided?

greglucas replied:
> Are you referring to the format of the ancillary files or a different file?

This file I commented on looks like this:
```
Species, Sector, Lower Energy (MeV), Upper Energy (MeV), Delta E (MeV), Geometry Factor (cm2 sr), Efficiency, b
H      , 0     , 1.8               , 4                 , 2.2          , 0.41686                 , 1         , 0
```

The next one down looks like this:
```
Species,Lower Energy (MeV),Upper Energy (MeV),Delta E (MeV),Geometry Factor (cm2 sr),Efficiency,b
H,1.8,2.2,0.4,3.4146,1,0
```

notice the spacing/formatting difference. The columns also have an extra space after the comma which will mess up pandas column names/access via string sometimes.

> Would mocking the tables be better than using the test ancillary files HIT provided?

IMO mocking is better for unit tests and making sure you have control over what you want to control. If you want to test whether your calibration file behaves differently for values outside of the 0-1 range, or with specific values you can multiply in a known way. These files are more "integration" tests, and are good for testing the actual data expected from the teams, but you have little control over them. So, both are valuable. I find the mock stuff much easier to keep moving fast with though because you aren't dependent on someone else.

vmartinez-cu replied:
> notice the spacing/formatting difference. The columns also have an extra space after the comma which will mess up pandas column names/access via string sometimes.

The first file you mentioned is for a different L2 product. I've removed it from this PR. To clean up formatting of the ancillary files, I have `ancillary_data.columns = ancillary_data.columns.str.lower().str.strip()`. This removes extra spaces around the column names and makes them lower case. This should hopefully address most formatting issues

greglucas commented on imap_processing/hit/l2/hit_l2.py:
```diff
@@ -0,0 +1,163 @@
+                dataset[field].attrs = attr_mgr.get_variable_attributes(field)
+                dataset[field].assign_coords(dims)
+            except KeyError:
+                print(f"Field {field} not found in attribute manager.")
+                logger.warning(f"Field {field} not found in attribute manager.")
```
It is OK to continue processing if this happens or do you want to re-raise?

You probably only need one or the other of the print / logger.warning

vmartinez-cu replied:
It should continue for now since I don't have L2 attributes defined yet and don't want processing to fail as a result. I'll add a note about this

greglucas reviewed (commented):
Looks great. A few minor questions mostly around the ancillary file handling.

tmplummer committed:
1402 hi l1c   pset get de clock ticks for esa step (#1404)

* Add get_de_clock_ticks_for_esa_step function

* Add test coverage for get_de_clock_ticks_for_esa_step function

vmartinez-cu committed:
Update imap_processing/cli.py

Fix typo

Co-authored-by: Tenzin Choedon <36522642+tech3371@users.noreply.github.com>

vmartinez-cu committed:
Update imap_processing/cli.py

Fix typo

Co-authored-by: Tenzin Choedon <36522642+tech3371@users.noreply.github.com>

vmartinez-cu committed:
Update imap_processing/cli.py

Fix typo

Co-authored-by: Tenzin Choedon <36522642+tech3371@users.noreply.github.com>

vmartinez-cu committed:
Update imap_processing/cli.py

Fix typo

Co-authored-by: Tenzin Choedon <36522642+tech3371@users.noreply.github.com>

tmplummer committed:
1416 hi l1a   decom diag fee packets (#1420)

* Add DIAG_FEE packet apids to HIAPID enum
Move setting of L1A global attributes to common location for all APIDs
Remove housekeeping module (no longer needed due to common gattr set location)
Add logic for processing DIAG_FEE packets

* Add test coverage and validation for hi diag_fee packets

* Add coverage for writing dataset to cdf

* Remove l1.housekeeping from doc build

* Make diagfee name consistent in CDF global attributes yaml

greglucas committed:
TST: Remove unused test files (#1422)

vmartinez-cu committed:
Remove ancillary files not related to this PR

nkerman committed:
"Pull" Index matching, value binning; Convert TT -> TDB (#1396)

vmartinez-cu committed:
Add TODO

vmartinez-cu committed:
Fix minor typo

greglucas committed:
ENH: XTCE: Add support for segmented polynomials (#1409)

* ENH: XTCE: Add support for segmented polynomials

The previous handling of segmented polynomials only considered the
first row and didn't use any of the other factors. This iterates
through the other rows creating additional calibrators and adds
a ContextCalibrator for each segment that checks against the low
and high value specified within the spreadsheet.

tmplummer committed:
Add auto generation of calibration product integer values column from… (#1408)

* Add auto generation of calibration product integer values column from detector strings

* Improve docstrings

* Move coincidence_type_str_to_in to a class method of CoincidenceBitmap
Move CoincidenceBitmap to the hi.uitils module
Move tests to appropriate files

greglucas committed:
MNT: Remove pyyaml from our direct dependencies (#1425)

tech3371 committed:
SWE: L1 and L2 CDF attrs fixes identified by SKTEditor (#1428)

tech3371 committed:
SWE: Add CDF attrs for non-science apid (#1430)

vmartinez-cu committed:
Use dictionary for L2 datasets rather than zipping a list of datasets with a list of logical sources for assigning attributes. Safer option
